### PR TITLE
#727 Remove SpriteAnimationComponent when animation is done

### LINKED ISSUE #727: SpriteAnimationComponent was removed prematurely
# Current bug behaviour

`SpriteAnimationComponent` was removed in the last frame immediately regardless of the `stepTime`.

```dart
@override
  bool get shouldRemove =>
      super.shouldRemove ||
      (removeOnFinish && (animation?.isLastFrame ?? false));
```

# Expected behaviour

I think `SpriteAnimationComponent` should be removed when `SpriteAnimation` is done.

```dart
@override
  bool get shouldRemove =>
      super.shouldRemove ||
      (removeOnFinish && (animation?.done() ?? false));
```

PS: `SpriteAnimationComponent` can never be automatically removed if `SpriteAnimation#loop` is true. Perhaps this is a breaking change.

# Steps to reproduce

# Flutter doctor output

```
Output of: flutter doctor -v
```

# More environment information

* Flame version: 1.0.0-rc8

# Log information

```
Enter log information in this code block
```

# More information

If my idea meets the needs, I'm interested in working on a PR for this.

### PR DESCRIPTION
**Pull request by wenxiangjiang**
# Description

We should only remove the `SpriteAnimationComponent` when the animation is really done, not when it is on the last frame.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flutter Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [ ] No, this is *not* a breaking change.

## Related Issues

fixes #727 

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

### CODE REVIEW

File: packages/flame/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@
  - Add clear function to BaseGame to allow the removal of all components
  - Moving tutorials to the Flame main repository
  - Exporting Images and AssetsCache
+ - Remove the SpriteAnimationComponent when the animation is really done, not when it is on the last frame (this is a breaking change)
```
Comment (wolfenrain) [inline packages/flame/CHANGELOG.md:10]: I know our PR states that you have to indicate a breaking change, but as this is for the v1 branch. Which is not officially released yet I dont think we need to indicate these breaking changes.

```suggestion
 - *Remove the SpriteAnimationComponent when the animation is really done, not when it is on the last frame
```

**Review by wolfenrain — APPROVED** [review]

File: packages/flame/CHANGELOG.md (hunk 1 above)
Comment (wenxiangjiang) [reply packages/flame/CHANGELOG.md]: I think so.

File: packages/flame/lib/src/components/sprite_animation_component.dart [hunk 2]
```diff
@@ -35,11 +35,12 @@ class SpriteAnimationComponent extends PositionComponent {
...
 
-  /// Component will be removed after loop end and [removeOnFinish] is set.
+  /// Component will be removed after animation is done and [removeOnFinish] is set.
+  ///
+  /// Note: [SpriteAnimationComponent] will not be removed automatically if loop property of [SpriteAnimation] is true.
   @override
   bool get shouldRemove =>
-      super.shouldRemove ||
-      (removeOnFinish && (animation?.isLastFrame ?? false));
+      super.shouldRemove || (removeOnFinish && (animation?.done() ?? false));
```
Comment (erickzanardo) [inline packages/flame/lib/src/components/sprite_animation_component.dart:43]: Would be nice to have an unit test here, to ensure this behaviour in the future

File: packages/flame/CHANGELOG.md [hunk 3]
```diff
@@ -7,7 +7,7 @@
  - Add clear function to BaseGame to allow the removal of all components
  - Moving tutorials to the Flame main repository
  - Exporting Images and AssetsCache
- - Remove the SpriteAnimationComponent when the animation is really done, not when it is on the last frame (this is a breaking change)
+ - *Remove the SpriteAnimationComponent when the animation is really done, not when it is on the last frame
```
Comment (wolfenrain) [inline packages/flame/CHANGELOG.md:10]:
```suggestion
 - Remove the SpriteAnimationComponent when the animation is really done, not when it is on the last frame
```

My mistake sorry.

**Review by wolfenrain — APPROVED** [review]

File: packages/flame/lib/src/components/sprite_animation_component.dart (hunk 2 above)
Comment (wenxiangjiang) [reply packages/flame/lib/src/components/sprite_animation_component.dart:43]: > Would be nice to have an unit test here, to ensure this behaviour in the future

I will try to add a unit test later.

**Review by wolfenrain — APPROVED** [review]

**Review by erickzanardo — APPROVED** [review]

**Review by spydon — APPROVED** [review]
Lgtm, great with that exhaustive test that you added!

File: packages/flame/test/components/sprite_animation_component_test.dart [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    data[i] = 255;
+    data[i + 1] = 255;
+    data[i + 2] = 255;
+    data[i + 3] = 255;
+  }
+  final image = await Flame.images.decodeImageFromPixels(
+    data,
+    1,
+    1,
+  );
```
Comment (luanpotter) [inline packages/flame/test/components/sprite_animation_component_test.dart:20]: nit: can this fit in one line? if not it's perfect like this

**Review by luanpotter — APPROVED** [review]
LGTM, great catch!

**Review by renancaraujo — APPROVED** [review]

**Pull request merged**